README.md
# crunchdev

[![Build Status](https://travis-ci.org/Crunch-io/crunchdev.png?branch=master)](https://travis-ci.org/Crunch-io/crunchdev)  [![codecov](https://codecov.io/gh/Crunch-io/crunchdev/branch/master/graph/badge.svg)](https://codecov.io/gh/Crunch-io/crunchdev)
[![cran](https://www.r-pkg.org/badges/version-last-release/crunchdev)](https://cran.r-project.org/package=crunchdev)

The crunchdev package offers a number of convenience functions and [RStudio addins](https://rstudio.github.io/rstudioaddins/) to make developing crunch packages quicker and faster. 

Highlights of CrunchDev include:
* *addins for quickly and easily specifying host, test type (unit vs integration), and file filters* These make it quick and easy to run tests on specific files which speeds up how quickly you can see if those test pass or fail, and can get back into the code. It handles all of the logic for setting up hosts and authentication for integration tests so you don't have to change any configurations or options to run them. Test are run in a terminal within RStudio so they won't block any other processes or coding you might want to do.
* *easily change backends interactively with `setupCrunch()`* There are a number of backends that are useful to use in the process of developing Crunch packages. This function lets you change which backend you are configured to connect to (eg changes the `crunch.email`, `crunch.pw`, and `crunch.api` options). No more fumbling with changing options or reloading R.

## Installing
The pre-release version of the package can be pulled from GitHub using the [devtools](https://github.com/hadley/devtools) package:

    # install.packages("devtools")
    devtools::install_github("Crunch-io/crunchdev")

## Configuration
Before using `crunchdev` there are a few options that should be added to your `~/.Rprofile`:

1. host name url maps --- there are two groups: `user.hosts` and `test.hosts` the first uses user authentication, and the second uses test user authentication. Each list can have any number of hostnames and urls.

```
crunchdev.user.hosts=c("stableapp" = "https://stableapp.crunch.io/api/"),
crunchdev.test.hosts=c(
       "local" = "http://local.crunch.io/api/",
       "testing server" = "https://testing.crunch.io/api/")
```
	       
1. authentication maps --- again, there are two groups: `user.auth` and `test.auth` the first is used for all hosts in `crunchdev.user.hosts`, and the second is used for all hosts in `crunchdev.test.hosts`. Each one should only contain one email and password.

```
crunchdev.user.auth=c("email" = "magic.user@crunch.io",
                      "pw" = "t0pSecretP@ssw0rD"),
crunchdev.test.auth=c("email" = "magic.testuser@crunch.io",
                      "pw" = "t0pSecretP@ssw0rD")
```
			  
### Example additions to an `.Rprofile`
```
options(
    crunchdev.user.auth=c("email" = "developer@crunch.io",
    "pw" = "super_secre7_passw0rd"),
    crunchdev.test.auth=c("email" = "tester@crunch.io",
    "pw" = "passw0rd_wh1ch_1s_super_secre7"),
    crunchdev.user.hosts=c("aleph" = "https://aleph.crunch.io/api/"),
    crunchdev.test.hosts=c(
	       "bet" = "https://bet.crunch.io/api/",
	       "gimmel" = "https://gimmel.crunch.io/api/")
)
```


## Addins
RStudio allows keyboard shortcuts to be bound to addins. The addins provided by crunchdev are designed to make iteration and test running quick and easy. *Note:* currently launching a terminal from RStudio requires installing [rstudioapi from feature/terminal](https://github.com/rstudio/rstudioapi/pull/52) and requires RStudio [version >= 1.1.331](https://dailies.rstudio.com/).

* Test running — For all of the test addins a configuration dialog will appear with optional settings. When done (or return) is pressed, a new terminal will launch and the tests will be run in a new, clean session.
  * Test all (Crunch) — test configuration preset to run all unit tests Suggest key binding: `shift+alt+u`
  * Test active doc (Crunch) — test configuration preset to run unit tests with the filter set for the active file (eg having `test-dataset-reference.R` active will prefill `dataset-reference` in the filter box) Suggest key binding: `shift+cmd+u`
  * Test active doc integration (Crunch) — test configuration preset to run unit and integration tests with the filter set for the active file Suggest key binding: `shift+cmd+i`
* Load all (Crunch) — tries to `logout()` before running `devtools::load_all()` this cleans up interactive iteration and reloading of Crunch packages, Since the `logout()` is run within `try()` this addin should be safe to remap `shift+cmd+l` globally.

## For developers

The repository includes a Makefile to facilitate some common tasks.

### Running tests

`$ make test`. Requires the [testthat](https://github.com/hadley/testthat) package. You can also specify a specific test file or files to run by adding a "file=" argument, like `$ make test file=logging`. `test_package` will do a regular-expression pattern match within the file names. See its documentation in the `testthat` package.

### Updating documentation

`$ make doc`. Requires the [roxygen2](https://github.com/klutometis/roxygen) package.


NEWS.md
## crunchdev 0.1.0

* Added `test_crunch()` brings up a shiny dialog with options to test
* Added `reload_test_data()` to quickly and easily reload testing data
* Added `setupCrunch()` to change auth and hostnames easily
* Initial addition of functions and tests


R/test_gadget.R
test_gadget <- function(filter, test_type = "unit") {
    hosts <- c(envOrOption("crunch.test.hosts"), envOrOption("crunch.user.hosts"))
    names(hosts) <- paste0(names(hosts), " - ", hosts)


    ui <- miniUI::miniPage(
        miniUI::gadgetTitleBar("Crunch test options"),
        miniUI::miniContentPanel(
            shiny::textInput("filter", "File filter:", filter),
            shiny::tags$script("document.getElementById('filter').focus()"), # focus on the filter
            shiny::tags$script('
                $(document).on("keypress", function (e) {
                   Shiny.onInputChange("keys", e.which);
                });
              ') ,
            shiny::radioButtons("type", "Test type:",
                         c("unit" = "unit",
                           "integration" = "integration"),
                         inline = TRUE,
                         test_type),
            # TODO: grey out host select if unit is selected
            shiny::selectInput("host", "Host:",
                                hosts)
        )
    )

    server <- function(input, output, session) {
        # Define reactive expressions, outputs, etc.

        # When the Done button is clicked, return a value
        shiny::observeEvent(input$done, {
            shiny::stopApp(return_values(input))
        })
        shiny::observeEvent(input$keys, {
            if (input$keys == 13) {
                # if return, return
                shiny::stopApp(return_values(input))
            }
        })


    }
    shiny::runGadget(ui, server, viewer = shiny::dialogViewer("Crunch test", width = 325, height = 375))
}

return_values <- function(input) {
    return(list(type=input$type, host=input$host, filter=input$filter))
}


tests/testthat.R
library(testthat)
test_check("crunchdev")


R/crunchdev.R
# Here's a good place to put your top-level package documentation

.onAttach <- function (lib, pkgname="crunchdev") {
    ## Put stuff here you want to run when your package is loaded
    invisible()
}

#' @import crunch
NULL

#' logout of Crunch before load_alling
#'
#' A replacement for [`devtools::load_all`] will try to logout before runing
#' `devtools::load_all` so that multiple logins are not in the same session
#'
#' @param ... options passed to `load_all`
#'
#' @return nothing
#'
#' @export
load_all <- function(...) {
    try(logout())
    devtools::load_all(...)
}


#' Function to configure and kick off tests of Crunch packages
#'
#' Gets document context, and then presents the test_gadget which has
#' configuration options for testing Crunch packages
#'
#' @param filter a character of the regex of the filter to pre-populate
#' the test gadget input with
#' @param ... options to pass to the test_gadget
#'
#' @return nothing
#'
#' @export
test_crunch <- function(filter = NULL, ...) {
    if (is.null(filter)) {
        # TODO: do some matching here to check if the active R/(.*).R doc has a
        # matching test-\\1.R, if not do all
        doc <- rstudioapi::getActiveDocumentContext()
        file <- basename(doc$path)
        filter <- gsub("^test-|\\.R$", "", file)
    } else {
        filter <- filter
    }

    test_opts <- test_gadget(filter = filter, ...)
    opts <- setup_host_auth(test_opts)

    test_cmd <- sprintf("R --slave -e 'library(httptest); options(crunch.check.updates=FALSE); devtools::test(filter=\"%s\")' \n", test_opts$filter)
    integration <- as.character(test_opts$type == "integration")
    crunch_terminal(
        test_cmd,
        env=c(R_TEST_API=opts$host, INTEGRATION=integration,
              R_TEST_USER=opts$user, R_TEST_PW=opts$pw))
}

test_crunch_int <- function() test_crunch(filter=NULL, test_type="integration")
test_crunch_all <- function() test_crunch(filter=".*")

setup_host_auth <- function (opts) {
    if (!grepl("https?://.*", opts$host)) {
        all_hosts <- c(envOrOption("crunchdev.test.hosts"),
                       envOrOption("crunchdev.user.hosts"))
        opts$host <- all_hosts[opts$host]
    }

    if (opts$host %in% envOrOption("crunchdev.test.hosts")) {
        opts$user <- envOrOption("crunchdev.test.auth")[["email"]]
        opts$pw <- envOrOption("crunchdev.test.auth")[["pw"]]
    } else {
        opts$user <- envOrOption("crunchdev.user.auth")[["email"]]
        opts$pw <- envOrOption("crunchdev.user.auth")[["pw"]]
    }
    return(opts)
}

#' Check package coverage and shine the result
#'
#' A convenience method for `covr::shine(covr::package_coverage())`
#'
#' @param ... options passed to `package_coverage`
#'
#' @return nothing
#'
#' @export
shine_covr <- function(...) {
    pkg_coverage <- covr::package_coverage(...)
    covr::shine(pkg_coverage)
}

#' Setup Crunch to use a particular backend
#'
#' If the host provided is in the `crunchdev.test.hosts` group, test authentication will be set (`crunchdev.test.auth`). Otherwise, user authentication will be set. If `host` is not recognized, it will be passed to [crunch::setCrunchAPI()]
#'
#' @param host host to connect to (e.g. "app", "local")
#' @param ... passed to [crunch::setCrunchAPI()]
#'
#' @return nothing
#'
#' @export
setupCrunch <- function (host, ...) {
    opts <- setup_host_auth(list("host" = host))

    if (is.na(opts$host)) {
        setCrunchAPI(host, ...)
    } else {
        options(
            "crunch.api" = opts$host
        )
    }

    options(
        "crunch.email" = opts$user,
        "crunch.pw" = opts$pw
    )

    message("Crunch will connect to ", getOption("crunch.api"),
            " with ", getOption("crunch.email"))

    return(invisible(opts))
}



R/crunch-terminal.R
crunch_terminal <- function (cmd, env, term_id = getOption("crunchdev.term"),
                             term_caption = "crunchdev") {
    # if there's no terminal id, create one
    if (is.null(term_id)) {
        # check if there are any extant term_caption terms
        for (t_id in rstudioapi::terminalList()) {
            cont <- rstudioapi::terminalContext(t_id)
            if (cont$caption == term_caption) {
                term_id <- t_id
            }
        }
    }

    if (is.null(term_id) || is.null(rstudioapi::terminalContext(term_id))) {
        # if there is no terminal, then create a new one
        term_id <- rstudioapi::terminalCreate(caption = term_caption)
    }

    # remember the terminal id
    options(crunchdev.term = term_id)


    rstudioapi::terminalActivate(term_id, show = TRUE)
    rstudioapi::terminalClear(term_id)

    # write, send, and remove env variables
    file <- write_env(env)
    rstudioapi::terminalSend(term_id, sprintf('source %s \n', file))
    rstudioapi::terminalSend(term_id, sprintf('rm %s \n', file))

    rstudioapi::terminalSend(term_id, cmd)
}

write_env <- function (env) {
    filename <- tempfile()
    envs <- make_env_strings(env)
    writeLines(envs, con=filename)
    return(filename)
}

make_env_strings <- function (env) {
    paste0('export ', names(env), "=", env, collapse='\n')
}



R/reload-test-data.R
#' Reload test data
#'
#' Reloads all types, mrdf, mrdf with a multiple resonse and apidocs test data
#' to whichever backend is set in options.
#'
#' @export
reload_test_data <- function () {
    login()

    message("Loading all types")
    ## Data frames to make datasets with
    df <- data.frame(v1=c(rep(NA_real_, 5), stats::rnorm(15)),
                     v2=c(letters[1:15], rep(NA_character_, 5)),
                     v3=8:27,
                     v4=as.factor(LETTERS[2:3]),
                     v5=as.Date(0:19, origin="1955-11-05"),
                     v6=TRUE,
                     stringsAsFactors=FALSE)
    newDataset(df, name = "test ds")

    mrdf <- data.frame(mr_1=c(1,0,1,NA_real_),
                       mr_2=c(0,0,1,NA_real_),
                       mr_3=c(0,0,1,NA_real_),
                       v4=as.factor(LETTERS[2:3]),
                       stringsAsFactors=FALSE)
    message("Loading cat array data")
    mrdf.setup(newDataset(mrdf, name = "test-mrdf"))
    message("Loading multiple response data")
    mrdf.setup(newDataset(mrdf, name = "test-mrdfmr"), selections = "1.0")
    message("Loading apidocs data")
    newDatasetFromFixture("./tests/testthat", "apidocs")

    return()
}

# copied directly from crunch, should be moved to crunchdev eventually
# setup a multiple response variable
mrdf.setup <- function (dataset, pattern="mr_", name=ifelse(is.null(selections),
                                                            "CA", "MR"), selections=NULL) {
    cast.these <- grep(pattern, names(dataset))
    dataset[cast.these] <- lapply(dataset[cast.these],
                                  crunch:::castVariable, "categorical")
    if (is.null(selections)) {
        dataset[[name]] <- makeArray(dataset[cast.these], name=name)
    } else {
        dataset[[name]] <- makeMR(dataset[cast.these], name=name,
                                  selections=selections)
    }
    return(dataset)
}


# copied directly from crunch, should be moved to crunchdev eventually
newDatasetFromFixture <- function (path_prefix, filename) {
    ## Grab csv and json from "dataset-fixtures" and make a dataset
    m <- crunch:::fromJSON(
        file.path(path_prefix, "dataset-fixtures", paste0(filename, ".json")),
        simplifyVector=FALSE)
    return(suppressMessages(
        createWithMetadataAndFile(
            m,
            file.path(path_prefix,"dataset-fixtures", paste0(filename, ".csv"))
            )))
}

R/test-gadget.R
test_gadget <- function(filter, test_type = "unit") {
    hosts <- c(envOrOption("crunchdev.test.hosts"), envOrOption("crunchdev.user.hosts"))
    names(hosts) <- paste0(names(hosts), " - ", hosts)


    ui <- miniUI::miniPage(
        miniUI::gadgetTitleBar("Crunch test options"),
        miniUI::miniContentPanel(
            shiny::textInput("filter", "File filter:", filter),
            shiny::tags$script("document.getElementById('filter').focus()"), # focus on the filter
            shiny::tags$script('
                $(document).on("keypress", function (e) {
                   Shiny.onInputChange("keys", e.which);
                });
              ') ,
            shiny::radioButtons("type", "Test type:",
                         c("unit" = "unit",
                           "integration" = "integration"),
                         inline = TRUE,
                         test_type),
            # TODO: grey out host select if unit is selected
            shiny::selectInput("host", "Host:",
                                hosts)
        )
    )

    server <- function(input, output, session) {
        # Define reactive expressions, outputs, etc.

        # When the Done button is clicked, return a value
        shiny::observeEvent(input$done, {
            shiny::stopApp(return_values(input))
        })
        shiny::observeEvent(input$keys, {
            if (input$keys == 13) {
                # if return, return
                shiny::stopApp(return_values(input))
            }
        })


    }
    shiny::runGadget(ui, server,
                     viewer = shiny::dialogViewer("Crunch test", width = 325, height = 375))
}

return_values <- function(input) {
    return(list(type=input$type, host=input$host, filter=input$filter))
}
